File: p19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # two pointers
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        temp = ListNode(0)
        # link temp to head
        # this deals with edge case
        temp.next = head

        front = temp
        back = temp

        for _ in range(n + 1):
            front = front.next

        while front is not None:
            front = front.next
            back = back.next

        back.next = back.next.next
        return temp.next

File: p19.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = new ListNode(0);
        temp->next = head;
        ListNode* front = temp;
        ListNode* back = temp;

        for (int _ = 0; _ < n; _ ++){
            front = front->next;
        }

        while (front->next != nullptr){
            front = front->next;
            back = back->next;
        }

        back->next = back->next->next;

        return temp->next;
    }
};